stan/math/rev/functor/root_finder.hpp

```diff
@@ -0,0 +1,101 @@
+ * @tparam FTuple A tuple holding functors whose signatures all match
+ *  `(GuessScalar g, Types&&... Args)`.
+ * @tparam GuessScalar Scalar type
+ * @tparam MinScalar Scalar type
+ * @tparam MaxScalar Scalar type
+ * @tparam Types Arg types to pass to functors in `f_tuple`
+ * @param f_tuple A tuple of functors to calculate the value and any derivates
+ * needed.
+ * @param guess An initial guess at the root value
+ * @param min The minimum possible value for the result, this is used as an
+ * initial lower bracket
+ * @param max The maximum possible value for the result, this is used as an
+ * initial upper bracket
+ * @param digits The desired number of binary digits precision
+ * @param max_iter An optional maximum number of iterations to perform. On exit,
+ * this is updated to the actual number of iterations performed
+ * @param args Parameter pack of arguments to pass the the functors in `f_tuple`
+ */
+template <
+    typename SolverFun, typename FTuple, typename GuessScalar,
+    typename MinScalar, typename MaxScalar, typename... Types,
+    require_any_st_var<GuessScalar, MinScalar, MaxScalar, Types...>* = nullptr,
+    require_all_stan_scalar_t<GuessScalar, MinScalar, MaxScalar>* = nullptr>
+auto root_finder_tol(SolverFun&& f_solver, FTuple&& f_tuple,
+                     const GuessScalar guess, const MinScalar min,
+                     const MaxScalar max, const int digits,
+                     std::uintmax_t& max_iter, Types&&... args) {
+  check_bounded("root_finder", "initial guess", guess, min, max);
+  check_positive("root_finder", "digits", digits);
+  check_positive("root_finder", "max_iter", max_iter);
+  auto arena_args_tuple
+      = make_chainable_ptr(std::make_tuple(eval(std::forward<Types>(args))...));
+  auto args_vals_tuple = apply(
+      [&](const auto&... args) {
+        return std::make_tuple(to_ref(value_of(args))...);
+      },
+      *arena_args_tuple);
+  // Solve the system
+  double theta_dbl = apply(
+      [&f_solver, &f_tuple, &max_iter, digits, guess_val = value_of(guess),
+       min_val = value_of(min), max_val = value_of(max)](auto&&... vals) {
+        return root_finder_tol(f_solver, f_tuple, guess_val, min_val, max_val,
+                               digits, max_iter, vals...);
+      },
+      args_vals_tuple);
+  double Jf_x;
+  auto&& f = std::get<0>(f_tuple);
+  {
+    nested_rev_autodiff nested;
+    stan::math::var x_var(theta_dbl);
+    stan::math::var fx_var = apply(
+        [&x_var, &f](auto&&... args) { return f(x_var, std::move(args)...); },
+        std::move(args_vals_tuple));
+    fx_var.grad();
+    Jf_x = x_var.adj();
+  }
+
+  /*
+   * Note: Because we put this on the callback stack, if `f` is a lambda
+   * its captures must be in Stan's arena memory or trivially destructable.
+   */
+  return make_callback_var(theta_dbl,
+                           [f, arena_args_tuple, Jf_x](auto& ret) mutable {
+                             // Eigen::VectorXd eta =
```

Comment:
hmmm... can you remind me what `ret` is?

Reply:
`ret` here is the return of the `root_finder_tol` function in reverse mode. Another way to think about this code is 

```cpp
var ret = theta_dbl;
reverse_pass_callback([f, arena_args_tuple, Jf_x, ret]() mutable {
                             {
                               nested_rev_autodiff rev;
                               double ret_val = ret.val();
                               auto x_nrad_ = apply(
                                   [&ret_val, &f](const auto&... args) {
                                     return eval(f(ret_val, args...));
                                   },
                                   *arena_args_tuple);
                               x_nrad_.adj() = -(ret.adj() / Jf_x);
                               grad();
                             }
                           });
return ret;
```